docs(linux): Security: Add SELinux documentation (#93)
TexasInstruments/processor-sdk-doc: source/linux/Foundational_Components/System_Security/SELinux.rst

Reviewer: For all console logs which are intended to target prefer using,
`root@<machine>:~#`

- Disabled mode is strongly discouraged; not only does the system avoid enforcing the SELinux policy, it also avoids labeling any persistent objects such as files, making it difficult to enable SELinux in the future.

.. note::

   This guide assumes that the user understands how to build the Linux Kernel and, Yocto or Debian distribution.

Filesystem Setup
----------------

Yocto
^^^^^

The following steps describe how to build SELinux user-space tools and configuration on Yocto. Please use :ref:`Processor SDK - Building the SDK with Yocto <building-the-sdk-with-yocto>` as reference.

1. On a fresh build, update the config file to use :file:`configs/arago-scarthgap-selinux-config.txt`. This config includes the opensource `meta-selinux` layer into the build and also adds additional features for the filesystem.

   .. note::

      The default policy is targeted. If you want to update it, add the variable ``PREFERRED_PROVIDER_virtual/refpolicy = "refpolicy-<your_choice>"`` to the :file:`local.conf` where ``<your_choice>`` is targeted, mls, mcs.

> 2. Build your custom image in build directory

Reviewer: ```suggestion
2. Build your :file:`tisdk-default-image` in build directory
```

This can't be called out as `custom image` since selinux features are only deployed in `tisdk-default-image` via dynamic layers - https://git.ti.com/cgit/arago-project/meta-arago/tree/meta-arago-distro/dynamic-layers/selinux/recipes-core/images?h=scarthgap

   .. code-block:: console

      $ MACHINE=<machine> bitbake -k tisdk-default-image

> 3. Flash the generated wic image onto your SD card. By default, SELinux functionality is in the disabled state

Reviewer: ```suggestion
3. Your newly built wic image will be generated in ``deploy-ti`` directory. Use :ref:`Linux SD Card Creation Guide <processor-sdk-linux-create-sd-card>` to flash this generated image onto your SD card. By default, SELinux functionality is in the disabled state
```

   .. code-block:: console

      $ sudo apt-get install selinux-basics selinux-policy-default auditd

Permissive mode Setup
---------------------

It is intended to go into Permissive mode, log the accesses, create the access policy and only then jump to Enforcing mode. Skipping this step would lead to a login failure due to lack of access.

1. Boot the target (in disabled state), Run ``fixfiles -F onboot`` or ``touch /.autorelabel`` to ensure that files are relabeled upon next reboot.

   .. warning::

      When systems run SELinux in permissive mode, users and processes might label various file-system objects incorrectly. File-system objects created while SELinux is disabled are not labeled at all. This behavior causes problems when changing to enforcing mode because SELinux relies on correct labels of file-system objects.

   To prevent incorrectly labeled and unlabeled files from causing problems, SELinux automatically relabels file systems when changing from the disabled state to permissive or enforcing mode.

2. Open :file:`/etc/selinux/config` and edit it to ``SELINUX=permissive`` and the policy type you are using.

> 

Author: I would not recommend adding any further information on this concept since we do not want to support queries about policy types. This guide seeks to explain how to enable SELinux, not serve as an introduction to the concept.

However, in future if we do decide to add it, I would also recommend adding in how to write a custom policy and fix policy errors.

> 

Author: I will move this to resolved for now, but you can bring it up again

> 3. Restart the system. If using U-Boot, add ``security=selinux`` to the boot parameters. Temporarily, you can stop at the U-Boot prompt and add,

Reviewer 2: what do we mean by "if using uboot". Is there another option for enabling se-linux if we are not using u-boot.

Author: This step only adds bootargs. Uboot is not enabling SELinux. There are other methods to pass bootargs to the kernel (via config or DT)

Author: However, I edited it to be a little clearer

   .. code-block:: console

>       setenv optargs "security=selinux"

Reviewer: ```suggestion
      => setenv optargs "security=selinux"
```

Similar to https://software-dl.ti.com/processor-sdk-linux/esd/AM62X/latest/exports/docs/linux/Foundational_Components/U-Boot/UG-General-Info.html#u-boot-environment

> .. _selinux_overview:

Author: Removed this altogther since no other references exist

- Enforcing mode is the recommended mode of operation where SELinux operates normally, enforcing the loaded security policy on the entire system.
- In permissive mode, the system acts as if SELinux is enforcing the loaded security policy, including labeling objects and emitting access denial entries in the logs, but it does not actually deny any operations. While not recommended for production systems, permissive mode can be helpful for SELinux policy development and debugging.
- Disabled mode is strongly discouraged; not only does the system avoid enforcing the SELinux policy, it also avoids labeling any persistent objects such as files, making it difficult to enable SELinux in the future.

.. note::

   This guide assumes that the user understands how to build the Linux Kernel and, Yocto or Debian distribution.

Filesystem Setup
----------------

Yocto
^^^^^

The following steps describe how to build SELinux user-space tools and configuration on Yocto. Please use :ref:`Processor SDK - Building the SDK with Yocto <building-the-sdk-with-yocto>` as reference.

1. On a fresh build, update the config file to use :file:`configs/arago-scarthgap-selinux-config.txt`.

   .. note::

>       The default policy is targeted. If you want to update it, add the variable PREFERRED_PROVIDER_virtual/refpolicy = "refpolicy-<your_choice>" where <your_choice> is targeted/mls/mcs in your local_conf file

Reviewer 3: ```suggestion
      The default policy is targeted. If you want to update it, add the variable ``PREFERRED_PROVIDER_virtual/refpolicy = "refpolicy-<your_choice>"`` to the :file:`local.conf`, where ``<your_choice>`` is targeted, mls, or mcs.
```

Debian
^^^^^^

The following steps describe how to install and configure SELinux together with the default policy which enables SELinux for the most important parts of your Debian filesystem. This guide does not describe how to build the filesystem using Debian.

1. The default Debian image provided by TI packages necessary kernel configs to enable SELinux. If you are not using the standard linux kernel provided by debian with an ext2/3/4/xfs/btrfs filesystem, you have to make sure that you are using a SELinux capable kernel and filesystem. For ext2/3/4 there is nothing special to be aware of, but for other filesystems there are still some quirks.

2. On the target machine, get the default policy and the basic set of SELinux utilities by running

   .. code-block:: console

      $ sudo apt-get install selinux-basics selinux-policy-default auditd

Permissive mode Setup
---------------------

It is intended to go into Permissive mode, log the accesses, create the access policy and only then jump to Enforcing mode. Skipping this step would lead to a login failure due to lack of access.

> 1. Boot the target (in disabled state), Run ``fixfiles -F onboot`` or ``touch /.autorelabel`` to configure GRUB and PAM and to create /.autorelabel.

Reviewer 3: ```suggestion
1. Boot the target (in disabled state), Run ``fixfiles -F onboot`` or ``touch /.autorelabel``
```

This command does not affect GRUB or PAM. Anything can create that file to trigger relabeling. If you mean that the user may use either of those tools to set that, that's also misleading. GRUB should not be modifying that file. There are different kernel parameters that should be used to control selinux instead.

Author: Thanks
